snitun/multiplexer/queue.py

```diff
@@ -0,0 +1,165 @@
+
+    def put_nowait(
+        self,
+        channel_id: MultiplexerChannelId,
+        message: MultiplexerMessage | None,
+    ) -> None:
+        """Put a message in the queue."""
+        self._put(channel_id, self._channels[channel_id], message)
+
+    def _put(
+        self,
+        channel_id: MultiplexerChannelId,
+        channel: _ChannelQueue,
+        message: MultiplexerMessage | None,
+    ) -> None:
+        """Put a message in the queue."""
+        size = 0 if message is None else len(message.data)
+        if channel.total_bytes >= self._channel_size_limit:
+            raise asyncio.QueueFull
+        channel.queue.append(message)
+        channel.total_bytes += size
+        self._order[channel_id] = None
+        self._wakeup_next(self._getters)
+
+    async def get(self) -> MultiplexerMessage | None:
+        """
+        Asynchronously retrieve a `MultiplexerMessage` from the queue.
+
+        Returns:
+            MultiplexerMessage: The message retrieved from the queue.
+
+        """
+        # Based on asyncio.Queue.get()
+        while not self._order:  # order is which channel_id to get next
+            getter = self._loop.create_future()
+            self._getters.append(getter)
+            try:
+                await getter
+            except:
+                getter.cancel()  # Just in case getter is not done yet.
+                with contextlib.suppress(ValueError):
+                    # Clean self._getters from canceled getters.
+                    self._getters.remove(getter)
+                # order is which channel_id to get next
+                if self._order and not getter.cancelled():
+                    # We were woken up by put_nowait(), but can't take
+                    # the call.  Wake up the next in line.
+                    self._wakeup_next(self._getters)
+                raise
+        return self.get_nowait()
+
+    def get_nowait(self) -> MultiplexerMessage | None:
+        """Get a message from the queue."""
+        if not self._order:
+            raise asyncio.QueueEmpty
+        channel_id, _ = self._order.popitem(last=False)
+        channel = self._channels[channel_id]
+        message = channel.queue.popleft()
+        size = 0 if message is None else len(message.data)
+        channel.total_bytes -= size
+        if channel.queue:
+            # Now put the channel_id back, but at the end of the queue
+            # so the next get will get the next waiting channel_id.
+            self._order[channel_id] = None
+        else:
+            # Got to the end of the queue
+            del self._channels[channel_id]
+        if putters := channel.putters:
+            self._wakeup_next(putters)
+        return message
+
+    def delete(self, channel_id: MultiplexerChannelId) -> None:
+        """Delete a channel queue."""
+        self._channels.pop(channel_id, None)
+        self._order.pop(channel_id, None)
+
+    def empty(self, channel_id: MultiplexerChannelId) -> bool:
```

Comment:
Channel_ prefixed methods for size/empty/full and than a size attribute (needs to be fast)? for the total size